File: 2020/Day11/Haskell/SeatingSystem-Part1.hs
import System.IO
import Control.Monad

printArray:: (Show a, Eq a) => [a] -> String
printArray (x:xs)
    | null xs = show x
    | otherwise = show x ++ "\n" ++ printArray xs

countNeighbours:: [[Char]] -> Int -> Int -> Char
countNeighbours world x y 
    | seat == '.' = '.'
    | seat == 'L' && adjacent == 0 = '#'
    | seat == '#' && adjacent >= 4 = 'L'
    | otherwise = seat
    where     
        seat = world!!y!!x
        lowerX = if x == 0 then x else x-1
        upperX = if x == length (head world) -1 then x else x+1
        lowerY = if y == 0 then y else y-1
        upperY = if y == length world-1 then y else y+1
        pos = [(i,j) | i <- [lowerX..upperX], j <- [lowerY..upperY], (x,y) /= (i, j)]
        adjacent = sum (map (\x -> fromEnum (world!!snd x!!fst x == '#')) pos)

solve:: [[Char]] -> Int -> [Int]
solve world previous
    | previous == current = [previous]
    | otherwise = current : solve newWorld current
    where 
        current = sum $ map (length . filter (=='#')) world
        newWorld = [[countNeighbours world x y | x <- [0..(length (head world)-1)]] | y <- [0..length world -1]]

main:: IO()
main = do
    handle <- openFile "../input.txt" ReadMode
    contents <- hGetContents handle
    let world = lines contents
    print (solve world 2)
    hClose handle

File: 2020/Day11/Python/SeatingSystem-Part1.py
import sys
import collections
import itertools
import numpy as np


def countNeigbours(world, x, y):
    if world[y, x] == ".":
        return "."
    lowerX = x if x == 0 else x-1
    upperX = x+1 if x == len(world[0])-1 else x+2
    lowerY = y if y == 0 else y-1
    upperY = y+1 if y == len(world)-1 else y+2
    seen = (world[lowerY:upperY, lowerX:upperX] == "#").sum() - (world[y][x] == "#")
    if world[y, x] == "L" and seen == 0:
        return "#"
    elif world[y, x] == "#" and seen >= 4:
        return "L"
    else:
        return world[y, x]

world = []

for line in open(f"{sys.path[0]}/../input.txt").read().splitlines():
    world.append(list(line))

world = np.array(world)
previous = -1
while previous != (world == "#").sum():
    previous = (world == "#").sum()
    print(previous)
    newWorld = np.array(world, copy=True)
    for y in range(len(world)):
        for x in range(len(world[y])):
            newWorld[y, x] = countNeigbours(world, x, y)
    world = np.array(newWorld)

print(previous)
